### ISSUE TITLE
[FEATURE] GitLab-Toolchain mwe

### ISSUE THREAD

**Issue by AMatutat** [PM-Dungeon/Lecture#46]
Wir brauchen ein MWE der angedachten Toolchain (Deploy to Abgabe). 
Für das MWE wäre Relevant
- läuft "über" einen Merge Request
- dummy JUnit-Test
- dummy SpotBugs
- dummy Formatierungs-Check
- die Ergebnisse der Toolchain werden im MR angezeigt und sind mindestens 120 Tage lang einsehbar.

**Comment by Tobias321** [comment]
Links:

- https://about.gitlab.com/devops-tools/github-vs-gitlab/ci-missing-github-capabilities/
- https://tomasvotruba.com/blog/how-can-we-use-github-actions-in-gitlab/

es scheint nicht unmöglich zu sein 😉

**Comment by cagix** [comment]
ja, aber immer im hinterkopf behalten, dass florian vermutlich noch dinge installieren muss (ich bin nicht sicher, ob die runner und docker schon in der basis-installation dabei sind) ...

**Comment by Tobias321** [comment]
Ja, die Runner und Docker sind noch nicht dabei, deshalb hatte ich das lokal gestartet. Florian müsste "Auto DevOps" hinzufügen.

**Comment by cagix** [comment]
schreib ihm bitte ne mail, cc an mich. (sicherheitshalber schreibe noch einen satz dazu, dass du unsere shk im projekt bist und die sache gerade organisierst).

**Comment by Tobias321** [comment]
Also diese Gradle-Tasks funktionieren:
- build/assemble
- junit test
- spotbugs test

Die GitHub-Action `axel-op/googlejavaformat-action` vom Marktplace könnte problematisch werden...

**Comment by Tobias321** [comment]
> schreib ihm bitte ne mail, cc an mich. (sicherheitshalber schreibe noch einen satz dazu, dass du unsere shk im projekt bist und die sache gerade organisierst).

📧 send 🙂

**Comment by cagix** [comment]
> Also diese Gradle-Tasks funktionieren:
> 
> * build/assemble
> * junit test
> * spotbugs test

👍 


>
> Die GitHub-Action `axel-op/googlejavaformat-action` vom Marktplace könnte problematisch werden...

ja, die wird es im gitlab nicht geben. 

aber wir hatten afaik irgendwann mal eine diskussion, ob das formatieren bei einem merge tatsächlich so eine gute idee ist und ob es nicht besser wäre, nur einen check mit checkstyle zu machen und das als vorbedingung für den merge mit aufzunehmen. 

das würde dann hier und "drüben" gehen. und für das formatieren wäre der einreicher selbst verantwortlich, wobei man die konfig vermutlich zur verfügung stellen müsste.

**Comment by Tobias321** [comment]
prinzipiell sind GitHub Actions ja auch nur bestimmte Scripts, die ausgeführt werden, allerdings müssen diese wohl auf Docker Hub verfügbar sein... (so hab ich jetzt jedenfalls den zweiten Link oben verstanden)

checkstyle hört sich gut an 👍🏻 nur, wie können wir das einbinden? (das jar File hinzufügen?) ... und wenn, dann müssen wir das ja konsistent/konsequent in beiden Repos machen, also hier und dort/drüben ;)

**Comment by cagix** [comment]
> prinzipiell sind GitHub Actions ja auch nur bestimmte Scripts, die ausgeführt werden, allerdings müssen diese wohl auf Docker Hub verfügbar sein... (so hab ich jetzt jedenfalls den zweiten Link oben verstanden)

hmmmja, oft sind das aber irgendwelche javascript-geschichten, die wieder x abhängigkeiten nach sich ziehen.

> 
> checkstyle hört sich gut an 👍🏻 nur, wie können wir das einbinden? (das jar File hinzufügen?) ... und wenn, dann müssen wir das ja konsistent/konsequent in beiden Repos machen, also hier und dort/drüben ;)

gab es nicht ein gradle-target, was das anstößt? und das jar wird dann vermutlich wie alle anderen auch als mvn-abhängigkeit formuliert?

**Comment by Tobias321** [comment]
Und Nelson ggf. auch.

> gab es nicht ein gradle-target, was das anstößt? und das jar wird dann vermutlich wie alle anderen auch als mvn-abhängigkeit formuliert?

Ich muss mir die Doku mal durchlesen...

**Comment by Tobias321** [comment]
Checkstyle funktioniert über ein Plugin, das in die `build.gradle` eingebunden wird, d. h., es gibt einen Gradle-Task, der aufgerufen werden kann (`gradlew check` ...).

Es gibt aber Probleme dabei. Die offizielle Google-Checkstyle-Konfiguration ist hier:
https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml

Es gibt aber keine offizielle AOSP-Style-Checkstyle-Konfiguration (sieh auch [hier](https://stackoverflow.com/questions/9339804/where-can-i-find-checkstyle-config-for-android-coding-style)).

Leider erhalte ich dadurch alleine bei zwei Java-Klassen schon gefühlt 100 Fehler. 🤷🏻

**Comment by cagix** [comment]
wo kommt denn der aktuell verwendete style her?

**Comment by Tobias321** [comment]
> wo kommt denn der aktuell verwendete style her?

Von diesem Link: https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml

das ist wohl die offizielle Checkstyle-Konfiguration von google... Die Datei hab ich nicht geändert.

Mit googlejavaformat verwenden wir ja bereits diesen Style, allerdings nicht in dem AOSP-Style (da gab es ja den Unterschied zwischen 2 und 4 Leerzeichen -Einrückung...).

**Comment by cagix** [comment]
> Mit googlejavaformat verwenden wir ja bereits diesen Style, allerdings nicht in dem AOSP-Style (da gab es ja den Unterschied zwischen 2 und 4 Leerzeichen -Einrückung...).

das war meine frage. welchen style verwendet das googlejavaformat wirklich? ist das wirklich https://google.github.io/styleguide/javaguide.html bzw. https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml? und wenn ja, warum wirft checkstyle dann so viele fehler, wenn du diesen style einstellst?

**Comment by Tobias321** [comment]
Ach so... wir verwenden google-java-format mit dem Code-Style Android Open Source Project:

![image](https://user-images.githubusercontent.com/85501570/151126038-87c32831-c320-4faf-8cd0-a0bb18eaee9e.png)

Und diesen Style habe ich bislang nicht für Checkstyle gefunden.

**Comment by cagix** [comment]
kann ij das ding evtl. exportieren? oder kann ij den style clonen und zum bearbeiten anbieten? dann wüsste man, was eingestellt ist.

**Comment by Tobias321** [comment]
Hab's herausgefunden, wir müssen nur
```
    <module name="Indentation">
      <property name="basicOffset" value="2"/>
      <property name="braceAdjustment" value="2"/>
      <property name="caseIndent" value="2"/>
      <property name="throwsIndent" value="4"/>
      <property name="lineWrappingIndentation" value="4"/>
      <property name="arrayInitIndent" value="2"/>
    </module>
```

ändern in:
```
    <module name="Indentation">
      <property name="basicOffset" value="4"/>
      <property name="braceAdjustment" value="4"/>
      <property name="caseIndent" value="4"/>
      <property name="throwsIndent" value="4"/>
      <property name="lineWrappingIndentation" value="4"/>
      <property name="arrayInitIndent" value="4"/>
    </module>
```

dann sind die Fehler weg, bis auf ein paar Java-Doc-Anmerkungen.

Siehe [auch](https://checkstyle.sourceforge.io/google_style.html).

**Comment by cagix** [comment]
sag ich doch :)
